### PR TITLE
[lint] add flake8-comment linter

### LINKED ISSUE #127352: `flake8: noqa` disables flake8 linter for the whole file and it's not obvious
For example,  test/dynamo/test_misc.py has `flake8: noqa` on line 7285, and it actually completely disables flake8 for the whole file.

Another example, torch/quantization/fx/utils.py has `flake8: noqa: F401`, but it actually again disables flake8 completely, not just rule F401.

Also I believe ruff respects the comment, so ruff is also disabled.

We should audit and fix the current usages, and maybe add a linter to complain about new `flake8: noqa` comments, as it masks real issues.

### PR DESCRIPTION
**Pull request by emmett-bicker**
Fixes #127352 

Adds a linter to check for `# flake8: noqa` comments that appear mid-file. It's implemented by using the built-in module tokenize to scan all comments and from those comments it identifies `# flake8: noqa` comments with a regex. It only flags errors after the first non-comment token (so the `# flake8: noqa` can only be preceded by other comments).

There are a considerable amount of files have mid-file  `# flake8: noqa` comments  which you can see by running
```bash
lintrunner lint --take FLAKE8_COMMENT --all-files
```

I wasn't sure how to deal with these preexisting mid-file comments, as adding in this linter would then make calls to `lintrunner --all-files` fail where it was previously successful, so I'd greatly appreciate any advice on that issue before the PR is considered for merging.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Comment by EmmettBicker** [comment]
This is also my first PR, so please let me know how I did :).

**Comment by kit1980** [comment]
Looks good overal.
I've started the CI, let's see what will fail (and tools/linter/adapters/flake8_comment_linter.py will probably fail itself).

**Comment by EmmettBicker** [comment]
@kit1980  Sounds great, thank you! I also wanted to ask if you had any advice for contributing more to PyTorch -- I plan to spend every day in the upcoming months contributing to PyTorch full time, and I'd love some guidance or direction in the process. The main thing I've done so far is try to contribute where I can and read Edward Yang's blog post titled [PyTorch Internals](http://blog.ezyang.com/2019/05/pytorch-internals/), but I'd really appreciate some general direction for issues I should look at!

**Comment by EmmettBicker** [comment]
@kit1980 Hi! Just checking in on this PR, does everything look good?

**Comment by kit1980** [comment]
@EmmettBicker You can see lots of lint errors everywhere: https://github.com/pytorch/pytorch/actions/runs/12127825244/job/33884448196?pr=141908

Those needs to be updated, either in this PR or separately (before merging this PR).

**Comment by EmmettBicker** [comment]
Thanks for letting me know! I was wondering what to do about all the errors this linter raises, and fixing them in a separate PR makes a lot of sense. I'll get on that!

File: torch/_logging/_internal.py
Comment (kit1980) [inline torch/_logging/_internal.py:647]: This should be probably just "noqa: B950".

"flake8: noqa: B950" actually disables everything flake8 for the whole file, and "noqa: B950" disables just rule B950 locally.

**Review by kit1980 — CHANGES_REQUESTED** [review]
I don't think it's OK to just move the existing noqa to the beginning.
The original intent was to have noqa just for some specific part in the middle - we don't want to lose that information.

Ideally each "flake8: noqa" in the middle should be replaced with a specific ignore, like "noqa: E731" (of even better resolved by changing the code).

At the very least, if there are too many violations, moving "flake8: noqa" to the top should be accompanied by the info from where it was moved (like function name) and probably a TODO to resolve.

File: benchmarks/dynamo/huggingface.py
Comment (kit1980) [inline benchmarks/dynamo/huggingface.py:50]: Move the comment also.

File: torch/_logging/_internal.py
Comment (EmmettBicker) [reply torch/_logging/_internal.py:647]: Thank you for the advice! I can certainly do that

**Comment by EmmettBicker** [comment]
> I don't think it's OK to just move the existing noqa to the beginning. The original intent was to have noqa just for some specific part in the middle - we don't want to lose that information.
> 
> Ideally each "flake8: noqa" in the middle should be replaced with a specific ignore, like "noqa: E731" (of even better resolved by changing the code).
> 
> At the very least, if there are too many violations, moving "flake8: noqa" to the top should be accompanied by the info from where it was moved (like function name) and probably a TODO to resolve.

That makes a lot of sense! I'll work on making more logical revisions to the existing code. 

In the suggested replacement given by the new linter, I made it suggest to move flake8: noqa to the top of the file -- do you think that's fine or should I come up with a different suggested replacement?

**Comment by kit1980** [comment]
> In the suggested replacement given by the new linter, I made it suggest to move flake8: noqa to the top of the file -- do you think that's fine or should I come up with a different suggested replacement?

I think it's fine. It will make clear to the author that flake8: noqa is for the whole file.

File: test/dynamo/test_misc.py
Comment (EmmettBicker) [inline test/dynamo/test_misc.py:7594]: @kit1980 I was going to fix this one by trying to disable something locally, but if I remove flake8: noqa for the whole file there are around 100 errors. I'd be happy to add in a few # noqa: specific errors to just ignore these, but it seems like a lot of shuffling code around. Which would you prefer I do: move the flake8: noqa to the top with a comment in this specific case, or add in specific # noqa: XXX around the file?

File: torch/_logging/_internal.py
Comment (EmmettBicker) [reply torch/_logging/_internal.py:647]: I actually don't get an error for this when I remove the noqa which means it's unnecessary -- can I just remove the noqa entirely?

Comment (kit1980) [reply torch/_logging/_internal.py:647]: Sure

**Review by kit1980 — APPROVED** [review]

**Comment by EmmettBicker** [comment]
Awesome! Thank you :)

**Comment by EmmettBicker** [comment]
@Chillee @titaiwangms @shubhambhokare1 @justinchuby @wschin Hi! If you all have time to look at this, I just moved around a few comments in code that you own.

**Comment by justinchuby** [comment]
The changes seem fine. Thanks!

**Comment by EmmettBicker** [comment]
I'm not sure if I'm supposed to do this, but

**Comment by EmmettBicker** [comment]
I also see that it's failing a lot of tests so that might not be great

**Comment by EmmettBicker** [comment]
Ok I'll figure out why there's so many fails

**Comment by EmmettBicker** [comment]
I think it failed this time because I tried before CI was green so I'm going to try again

**Comment by kit1980** [comment]
@EmmettBicker As you don't have repo write permissions, for security reasons someone needs to approve CI run after every change. I've just approved.

**Comment by EmmettBicker** [comment]
Ok thank you so much! That makes sense

**Comment by EmmettBicker** [comment]
I did not mean to delete that! Well I did but I didn't mean to remove all progress on this PR, I was confused about how to fix the linter issues so I put it on hold for a bit, but will likely return soon. I had removed the fork to refork with linux support.

**Pull request closed without merging**
